### PR TITLE
Visualize Lidar Plugin for ign-gazebo

### PR DESCRIPTION
**Pull request by mihirk284**
@ahcorde Can you please review it.

This PR integrates the LidarVisual from ignition rendering into ign-gazebo. Options are provided to select the visual type, the subscribed topic and to display or ignore lidar rays that do not hit an obstacle.

Example1

### CODE REVIEW

File: src/gui/plugins/visualize_lidar/VisualizeLidar.hh [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    public: Q_INVOKABLE QString TopicName() const;
+
+    /// \brief Set whether to display non-hitting rays
+    /// \param[in] _value Boolean value for displaying non hitting rays
+    public slots: void UpdateNonHitting(bool _value);
+
+    /// \brief Notify that topic name has changed
+    signals: void TopicNameChanged();
+
+    /// \brief Notify that topic name has changed
```
Comment (ahcorde) [inline src/gui/plugins/visualize_lidar/VisualizeLidar.hh:111]: review the documentation here and below

File: examples/worlds/visualize_lidar.sdf [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+              <min>0.08</min>
+              <max>10.0</max>
+              <resolution>0.01</resolution>
+            </range>
+          </lidar>
+          <alwaysOn>1</alwaysOn>
+          <visualize>true</visualize>
+        </sensor>
+      </link>
+
```
Comment (ahcorde) [inline examples/worlds/visualize_lidar.sdf:236]: remove white line

File: examples/worlds/visualize_lidar.sdf [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+      <static>true</static>
+    </model>
+
+    <model name='vehicle_blue'>
+      <pose>-4 0 0.325 0 0 0.0</pose>
+
```
Comment (ahcorde) [inline examples/worlds/visualize_lidar.sdf:242]: remove white line

File: examples/worlds/visualize_lidar.sdf [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+        <odom_publish_frequency>1</odom_publish_frequency>
+      </plugin>
+
+    </model>
+
+    <include>
+      <pose>0 0 0 0 0 1.57</pose>
+      <uri>
+      https://fuel.ignitionrobotics.org/1.0/openrobotics/models/Playground
+      </uri>
```
Comment (ahcorde) [inline examples/worlds/visualize_lidar.sdf:483]:
```suggestion
      <uri>https://fuel.ignitionrobotics.org/1.0/openrobotics/models/Playground</uri>
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: src/gui/plugins/visualize_lidar/VisualizeLidar.cc [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+#include <sdf/Link.hh>
+#include <sdf/Model.hh>
+
+#include <ignition/common/Animation.hh>
+#include <ignition/common/Console.hh>
+#include <ignition/common/KeyFrame.hh>
+#include <ignition/common/MeshManager.hh>
+#include <ignition/common/Profiler.hh>
+#include <ignition/common/Uuid.hh>
+#include <ignition/common/VideoEncoder.hh>
```
Comment (iche033) [inline src/gui/plugins/visualize_lidar/VisualizeLidar.cc:30]: most of these headers are not needed. Can you remove the ones not used?

File: src/gui/plugins/visualize_lidar/VisualizeLidar.cc [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+    public: std::string topicName;
+
+    /// \brief Entity representing the sensor in the world
+    public: gazebo::Entity lidarEntity;
+
+    /// \brief Minimum range for the visual
+    public: double minVisualRange;
+
+    /// \brief Maximum range for the visual
+    public: double maxVisualRange;
```
Comment (iche033) [inline src/gui/plugins/visualize_lidar/VisualizeLidar.cc:109]: initialize the min max range variables to 0.0

File: src/gui/plugins/visualize_lidar/VisualizeLidar.cc [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+  {
+    this->dataPtr->lidarPose = worldPose(this->dataPtr->lidarEntity, _ecm);
+  }
+}
+
+//////////////////////////////////////////////////
+void VisualizeLidar::UpdateType(int _type)
+{
+  std::lock_guard<std::mutex> lock(this->dataPtr->serviceMutex);
+  switch (_type) {
```
Comment (iche033) [inline src/gui/plugins/visualize_lidar/VisualizeLidar.cc:323]: nit: bracket on new line

File: src/gui/plugins/visualize_lidar/VisualizeLidar.cc [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+    case 0: this->dataPtr->visualType =
+                      rendering::LidarVisualType::LVT_NONE;
+            break;
```
Comment (iche033) [inline src/gui/plugins/visualize_lidar/VisualizeLidar.cc:326]: `this->dataPtr->visualType =` on new line with two space indentation

File: src/gui/plugins/visualize_lidar/VisualizeLidar.cc [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+                common::trimmed(data_values.value(0))) != 0)
+        {
+          this->dataPtr->lidarString = common::trimmed(data_values.value(0));
+          this->dataPtr->lidarEntityDirty = true;
+          this->dataPtr->maxVisualRange = this->dataPtr->msg.range_max();
+          this->dataPtr->minVisualRange = this->dataPtr->msg.range_min();
+          this->dataPtr->lidar->SetMaxRange(this->dataPtr->maxVisualRange);
+          this->dataPtr->lidar->SetMinRange(this->dataPtr->minVisualRange);
+          this->MinRangeChanged();
+          this->MaxRangeChanged();
```
Comment (iche033) [inline src/gui/plugins/visualize_lidar/VisualizeLidar.cc:418]: `break` here?

File: src/gui/plugins/visualize_lidar/VisualizeLidar.cc [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+            }
+          }
+          if (foundChild == false)
+          {
+            ignerr << "The entity could not be found."
+                  << "Error displaying lidar visual" <<std::endl;
+            return;
+          }
+        }
+        if (success == true)
```
Comment (iche033) [inline src/gui/plugins/visualize_lidar/VisualizeLidar.cc:305]: `if (success)`

File: src/gui/plugins/visualize_lidar/VisualizeLidar.cc [hunk 11]
```diff
@@ -0,0 +1,445 @@
...
+              parent = child;
+              foundChild = true;
+              if (i+1 == lidarURIVec.size()-1)
+              {
+                success = true;
+              }
+              break;
+            }
+          }
+          if (foundChild == false)
```
Comment (iche033) [inline src/gui/plugins/visualize_lidar/VisualizeLidar.cc:298]: `if (!foundChild)`

File: src/gui/plugins/visualize_lidar/VisualizeLidar.cc [hunk 12]
```diff
@@ -0,0 +1,445 @@
...
+        ignerr << "Error entity " << lidarURIVec[0]
+            << " doesn't exist and cannot be used to set lidar visual pose"
+            << std::endl;
+        return;
+      }
+      else
+      {
+        auto parent = baseEntity;
+        bool success = false;
+        for (auto i = 0ul; i < lidarURIVec.size()-1; i++)
```
Comment (iche033) [inline src/gui/plugins/visualize_lidar/VisualizeLidar.cc:277]: `auto i = 0ul` -> `size_t i = 0u`

File: src/gui/plugins/visualize_lidar/VisualizeLidar.cc [hunk 13]
```diff
@@ -0,0 +1,445 @@
...
+          for (auto child : children)
+          {
+            std::string nextstring = lidarURIVec[i+1];
+            std::string childname = std::string(
+                            _ecm.Component<components::Name>(child)->Data());
+            if (nextstring.compare(childname) == 0)
+            {
+              parent = child;
+              foundChild = true;
+              if (i+1 == lidarURIVec.size()-1)
```
Comment (iche033) [inline src/gui/plugins/visualize_lidar/VisualizeLidar.cc:291]: I think the logic in this block assumes the model > link > sensor hierarchy? 

We also support nested models, i.e.g model > model > link > sensor, in which case the logic above may not work?

File: src/gui/plugins/visualize_lidar/VisualizeLidar.cc [hunk 14]
```diff
@@ -0,0 +1,445 @@
...
+  }
+
+  // Create lidar visual
+  igndbg << "Creating lidar visual" << std::endl;
+
+  auto root = scene->RootVisual();
+  this->dataPtr->lidar = scene->CreateLidarVisual();
+  if (!this->dataPtr->lidar)
+  {
+    ignwarn << "Failed to create lidar, lidar visual plugin won't work."
```
Comment (iche033) [inline src/gui/plugins/visualize_lidar/VisualizeLidar.cc:193]: `lidar visual plugin` -> `visualize lidar plugin`?

File: src/gui/plugins/visualize_lidar/VisualizeLidar.cc [hunk 15]
```diff
@@ -0,0 +1,445 @@
...
+  if (!scene->IsInitialized() || scene->VisualCount() == 0)
+  {
+    return;
+  }
+
+  // Create lidar visual
+  igndbg << "Creating lidar visual" << std::endl;
+
+  auto root = scene->RootVisual();
+  this->dataPtr->lidar = scene->CreateLidarVisual();
```
Comment (iche033) [inline src/gui/plugins/visualize_lidar/VisualizeLidar.cc:190]: I don't see the visual destroyed? we need to be careful not to leave any dangling visuals around when the plugin exits

File: src/gui/plugins/visualize_lidar/VisualizeLidar.cc [hunk 16]
```diff
@@ -0,0 +1,524 @@
...
+}
+
+//////////////////////////////////////////////////
+void VisualizeLidar::DisplayVisual(bool _value)
+{
+  std::lock_guard<std::mutex>(this->dataPtr->serviceMutex);
+  if (!_value)
+  {
+    ignmsg << "Lidar Visual Display OFF." << std::endl;
+    this->dataPtr->lidar->SetType(rendering::LidarVisualType::LVT_NONE);
```
Comment (iche033) [inline src/gui/plugins/visualize_lidar/VisualizeLidar.cc:407]: you should be able to call `this->dataPtr->lidar->SetVisible(false)` to make it invisible, which is a faster operation than clearing all the lidar data. Same for re-enabling the lidar visual display below

File: src/gui/plugins/visualize_lidar/VisualizeLidar.cc (hunk 15 above)
Comment (iche033) [reply src/gui/plugins/visualize_lidar/VisualizeLidar.cc:192]: following up on this, we should destroy the visual when the user closes the plugin, i.e.  by clicking on `X` in the corner of the plugin window. Currently I see that lidar visual remains in the scene after exiting the plugin, and when I open the same plugin again, a duplicate lidar visual gets created.

**Comment by iche033** [comment]
one small issue: 
1.  uncheck `Display Lidar Visual` -> lidar becomes invisible - correct
1. uncheck `Show Non Hitting Rays` -> lidar becomes visible again when it should stay invisible

**Comment by mihirk284** [comment]
> one small issue:
> 
> 1. uncheck `Display Lidar Visual` -> lidar becomes invisible - correct
> 2. uncheck `Show Non Hitting Rays` -> lidar becomes visible again when it should stay invisible

@iche033  Thanks for bringing it to my attention. I think this is happening whenever the `DynamicRenderables` inside the visual are being deleted and new renderables are being used to display the visual. This happens in the case when there is a change in the `LidarVisualType` or if we have `displayNonHitting = false`.
This is happening despite having `lidar->SetVisible(false)`. The new `DynamicRenderables` generated after this operation always cause the visual to show up.

I can solve this by not updating the visual in the plugin if the display option is off.
Or should we change something in the rendering part of the code?

**Comment by iche033** [comment]
> Or should we change something in the rendering part of the code?
sounds like once we clear the visual data and create new ones in ign-rendering, we need to call `SetVisible` again with the current visible value so that the visibility is correct.

**Comment by iche033** [comment]
> sounds like once we clear the visual data and create new ones in ign-rendering, we need to call SetVisible again with the current visible value so that the visibility is correct.

fixed in ignitionrobotics/ign-rendering#133

**Comment by iche033** [comment]
changes look good. Can you merge with / rebase on `master`?

**Review by iche033 — APPROVED** [review]

File: src/gui/plugins/visualize_lidar/VisualizeLidar.cc [hunk 17]
```diff
@@ -0,0 +1,516 @@
...
+  if (!this->dataPtr->lidar)
+  {
+    ignwarn << "Failed to create lidar, visualize lidar plugin won't work."
+            << std::endl;
+
+    scene->DestroyVisual(this->dataPtr->lidar);
+
+    ignition::gui::App()->findChild<
+        ignition::gui::MainWindow *>()->removeEventFilter(this);
+    return;
```
Comment (ahcorde) [inline src/gui/plugins/visualize_lidar/VisualizeLidar.cc:200]: you can remove this line, right?

**Review by ahcorde — APPROVED** [review]
LGTM, two small comments

**Review by chapulina — COMMENTED** [review]
The Mac build failed, I think we need to bump the `ign-rendering` revision on `homebrew-simulation`:

```
/Users/jenkins/workspace/ignition_gazebo-ci-pr_any-homebrew-amd64/ign-gazebo/src/gui/plugins/visualize_lidar/VisualizeLidar.cc:49:10: fatal error: 'ignition/rendering/LidarVisual.hh' file not found
#include "ignition/rendering/LidarVisual.hh"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by chapulina** [comment]
@osrf-jenkins run tests, let's see what happens now

**Comment by chapulina** [comment]
@osrf-jenkins run tests :crossed_fingers:

**Pull request merged**